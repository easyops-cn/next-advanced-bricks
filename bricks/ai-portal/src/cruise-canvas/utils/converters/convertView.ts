import { pipes } from "@easyops-cn/brick-next-pipes";
import type {
  Component,
  ConvertViewOptions,
  DataSource,
  Variable,
  ViewWithInfo,
} from "./interfaces.js";
import convertList from "./convertList.js";
import convertDescriptions from "./convertDescriptions.js";
import convertDashboard from "./convertDashboard.js";
import convertTable from "./convertTable.js";
import type { BrickConf, ContextConf } from "@next-core/types";
import convertButton from "./convertButton.js";
import convertForm from "./convertForm.js";
import convertFlexLayout from "./convertFlexLayout.js";
import convertFormItem from "./convertFormItem.js";
import convertModal from "./convertModal.js";
import { convertEvents } from "./convertEvents.js";
import { withBox } from "./withBox.js";

export interface BrickConfWithContext extends BrickConf {
  context?: ContextConf[];
}

export async function convertView(
  view: ViewWithInfo | null | undefined,
  options: ConvertViewOptions
): Promise<BrickConfWithContext | null> {
  if (!view) {
    return null;
  }

  const componentMap = new Map<string, Component>();
  const childrenMap = new Map<string, Component[]>();
  const rootComponents: Component[] = [];
  const convertedComponents = new Set<string>();

  for (const component of view.components) {
    componentMap.set(component.componentId, component);
    if (component.parentComponentId) {
      let children = childrenMap.get(component.parentComponentId);
      if (!children) {
        children = [];
        childrenMap.set(component.parentComponentId, children);
      }
      children.push(component);
    } else {
      rootComponents.push(component);
    }
  }

  const convert = async (component: Component): Promise<BrickConf | null> => {
    if (convertedComponents.has(component.componentId)) {
      throw new Error(
        `Component ${component.componentId} has been converted already, it indicates a circular component tree.`
      );
    }
    convertedComponents.add(component.componentId);

    let brick: BrickConf | null = null;
    switch (component.componentName) {
      case "list":
        brick = await convertList(component);
        break;
      case "table":
        brick = await convertTable(component, view, options);
        break;
      case "descriptions":
        brick = await convertDescriptions(component, view, options);
        break;
      case "dashboard":
        brick = await convertDashboard(component, view, options);
        break;
      case "button":
        brick = await convertButton(component);
        break;
      case "form":
        brick = await convertForm(component);
        break;
      case "form-item":
        brick = await convertFormItem(component);
        break;
      case "flex-layout":
        brick = await convertFlexLayout(component);
        break;
      case "modal":
        brick = await convertModal(component);
        break;
    }

    if (!brick) {
      return null;
    }

    brick.events = convertEvents(component, view, options);

    // Set [data-component-id] for the brick
    brick.properties ??= {};
    brick.properties.dataset ??= {};
    (brick.properties.dataset as Record<string, string>).componentId =
      component.componentId;

    const children = (
      await Promise.all(
        (childrenMap.get(component.componentId) ?? []).map(convert)
      )
    ).filter(Boolean) as BrickConf[];

    if (children.length > 0) {
      brick.children = [...(brick.children ?? []), ...children];
    }

    return brick;
  };

  const context = [
    ...convertDataSourcesToContext(view.dataSources ?? []),
    ...convertVariablesToContext(view.variables ?? []),
    {
      name: "__builtin_fn_getLatestMetricValue",
      value: function getLatestMetricValue(
        data: {
          list?: Record<string, any>[];
        },
        {
          metric,
          precision,
        }: { metric: { id: string; unit: string }; precision?: number }
      ) {
        const value = data?.list?.findLast?.(
          (item) => item[metric.id] != null
        )?.[metric.id];
        const unit = metric.unit === "load" ? "" : metric.unit;
        return pipes.unitFormat(value, unit, precision).join("");
      },
    },
    {
      name: "SIZE",
      value: options.expanded ? "medium" : "small",
    },
  ];

  const children = (await Promise.all(rootComponents.map(convert))).filter(
    Boolean
  ) as BrickConf[];

  const needBox = rootComponents.every((component) =>
    ["form", "descriptions", "button"].includes(component.componentName)
  );

  return {
    brick: "eo-content-layout",
    children: needBox ? [withBox(children, options)] : children,
    context,
  };
}

function convertDataSourcesToContext(dataSources: DataSource[]): ContextConf[] {
  return dataSources.map((dataSource) => ({
    name: dataSource.name,
    resolve: {
      useProvider: `${dataSource.api.name}:${dataSource.api.version}`,
      params: dataSource.params,
      ...(dataSource.transform
        ? { transform: { value: dataSource.transform } }
        : null),
    },
    track: true,
  }));
}

function convertVariablesToContext(variables: Variable[]): ContextConf[] {
  return variables.map(({ name, value }) => ({
    name,
    value,
  }));
}
