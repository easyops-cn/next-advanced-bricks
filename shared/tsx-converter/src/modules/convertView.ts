import type {
  BrickConfInTemplate,
  CustomTemplate,
  StoryboardFunction,
} from "@next-core/types";
import type { ParsedApp } from "@next-shared/tsx-parser";
import convertModule, {
  type ConvertedPartOfComponent,
} from "./convertModule.js";
import type {
  ConvertState,
  ConvertOptions,
  ConvertResult,
} from "../interfaces.js";
import { getHelperFunctions } from "../helpers/index.js";
import { getViewTplName } from "./getTplName.js";
import { withBox } from "../withBox.js";

export async function convertView(
  view: ParsedApp,
  options: ConvertOptions
): Promise<ConvertResult> {
  const { entry } = view;
  if (!entry) {
    throw new Error("No entry module found in the view.");
  }

  const state: ConvertState = {
    usedHelpers: new Set(),
    app: view,
  };
  const convertedEntry = await convertModule(entry, state, options);

  const { defaultExport, internals, namedExports } = convertedEntry;

  if (!defaultExport) {
    throw new Error("No view found in the entry module.");
  }

  const functions: StoryboardFunction[] = [];
  const templates: CustomTemplate[] = [];

  for (const part of [...internals, ...namedExports.values()]) {
    if (part.type === "function") {
      functions.push(part.function);
    } else if (part.type === "template") {
      templates.push({
        name: getViewTplName(part.name!, options.rootId),
        bricks: part.bricks as BrickConfInTemplate[],
        state: part.context,
      });
    }
  }

  const helpers = Array.from(state.usedHelpers).map<StoryboardFunction>(
    (name) => {
      const source = getHelperFunctions().get(name);
      if (!source) {
        throw new Error(`Helper function ${name} not found`);
      }
      return {
        name,
        source,
        typescript: true,
      };
    }
  );

  const { title, bricks, context } = defaultExport as ConvertedPartOfComponent;

  const needBox = bricks.every((brick) =>
    ["eo-form", "eo-descriptions", "eo-button"].includes(brick.brick)
  );

  return {
    title,
    brick: {
      brick: "eo-content-layout",
      children: needBox ? [withBox(bricks, options)] : bricks,
    },
    context: [
      ...context,
      ...(options.withContexts
        ? Object.entries(options.withContexts).map(([name, value]) => ({
            name,
            value,
          }))
        : []),
    ],
    functions: [...functions, ...helpers],
    templates,
  };
}
