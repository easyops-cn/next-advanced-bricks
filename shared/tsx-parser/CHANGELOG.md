# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.1.0](https://github.com/easyops-cn/next-advanced-bricks/compare/@next-shared/tsx-parser@0.0.4...@next-shared/tsx-parser@0.1.0) (2025-09-17)


### Bug Fixes

* bailout earlier ([15cafbb](https://github.com/easyops-cn/next-advanced-bricks/commit/15cafbb8494aa13d3a7e56c542ac8747b4bd6a78))
* useState instead of createSignal ([8edb705](https://github.com/easyops-cn/next-advanced-bricks/commit/8edb7059d061ea66c527c7dba6d780e5bd7973a2))


### Features

* support function components ([e1025fd](https://github.com/easyops-cn/next-advanced-bricks/commit/e1025fd714c8c9b0d9e54a2c430e184ab9dff283))
* support useResource ([35d9c0d](https://github.com/easyops-cn/next-advanced-bricks/commit/35d9c0dc998667be5ed27e85a4789a39e292f9a4))





## [0.0.4](https://github.com/easyops-cn/next-advanced-bricks/compare/@next-shared/tsx-parser@0.0.3...@next-shared/tsx-parser@0.0.4) (2025-09-15)


### Bug Fixes

* fix a typo ([e8ede78](https://github.com/easyops-cn/next-advanced-bricks/commit/e8ede789005b3d0f47700f0b16056eed4e9a3b64))





## [0.0.3](https://github.com/easyops-cn/next-advanced-bricks/compare/@next-shared/tsx-parser@0.0.2...@next-shared/tsx-parser@0.0.3) (2025-09-12)


### Bug Fixes

* refine tsx view, support functions ([198d142](https://github.com/easyops-cn/next-advanced-bricks/commit/198d14281d282e5717745b74ad8b66b333ba1616))





## [0.0.2](https://github.com/easyops-cn/next-advanced-bricks/compare/@next-shared/tsx-parser@0.0.1...@next-shared/tsx-parser@0.0.2) (2025-09-11)


### Bug Fixes

* remove type annotations in expressions ([8b8e2a6](https://github.com/easyops-cn/next-advanced-bricks/commit/8b8e2a654b0944f18c4df9576a32680712bfa163))





## 0.0.1 (2025-09-10)

**Note:** Version bump only for package @next-shared/tsx-parser
