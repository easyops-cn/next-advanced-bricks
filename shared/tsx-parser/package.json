{
  "name": "@next-shared/tsx-parser",
  "version": "0.6.1",
  "homepage": "https://github.com/easyops-cn/next-advanced-bricks/tree/master/shared/tsx-parser",
  "repository": {
    "type": "git",
    "url": "git@github.com:easyops-cn/next-advanced-bricks.git",
    "directory": "shared/tsx-parser"
  },
  "license": "GPL-3.0",
  "sideEffects": false,
  "type": "module",
  "files": [
    "dist",
    "lib"
  ],
  "exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./lib/*": "./lib/*",
    "./package.json": "./package.json"
  },
  "scripts": {
    "clean": "rimraf dist",
    "prestart": "npm run clean",
    "start": "concurrently -k -n types,main \"npm run start:types\" \"npm run start:main\"",
    "start:main": "cross-env NODE_ENV=development build-next-libs --watch",
    "start:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types --project tsconfig.build.json --watch",
    "prebuild": "npm run clean",
    "build": "npm run build:types && npm run build:main",
    "build:main": "cross-env NODE_ENV=production build-next-libs",
    "build:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types --project tsconfig.build.json",
    "test": "cross-env NODE_ENV='test' test-next",
    "test:ci": "cross-env NODE_ENV='test' CI=true test-next"
  },
  "dependencies": {
    "@babel/parser": "^7.27.2",
    "@babel/traverse": "^7.27.4",
    "@babel/types": "^7.27.7",
    "@next-core/types": "^1.18.0"
  },
  "devDependencies": {
    "@next-core/build-next-libs": "^1.0.25",
    "@next-core/test-next": "^2.0.1"
  }
}
