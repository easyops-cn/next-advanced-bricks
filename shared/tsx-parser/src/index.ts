export { parseTsx } from "./parseTsx.js";
export * from "./interfaces.js";

export { isExpressionString } from "./utils.js";
