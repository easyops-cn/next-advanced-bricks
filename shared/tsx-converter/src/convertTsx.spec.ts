import { readFile } from "node:fs/promises";
import path from "node:path";
import { parseView } from "@next-shared/tsx-parser";
import { convertView } from "./modules/convertView.js";

describe("convertTsx", () => {
  test("should work", async () => {
    const code = await readFile(
      path.join(__dirname, "../../tsx-parser/src/__fixtures__/logs.txt"),
      "utf-8"
    );

    const view = parseView(code);
    const result = await convertView(view, { rootId: "test-root" });
    expect(result).toMatchSnapshot();
  });
});
