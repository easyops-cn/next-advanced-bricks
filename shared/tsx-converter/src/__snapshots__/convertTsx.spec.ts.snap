// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`convertTsx should work 1`] = `
{
  "brick": {
    "brick": "eo-content-layout",
    "children": [
      {
        "brick": "eo-flex-layout",
        "children": [
          {
            "brick": "eo-search",
            "events": {
              "search": [
                {
                  "else": [
                    {
                      "action": "context.replace",
                      "args": [
                        "logId",
                        undefined,
                      ],
                    },
                    {
                      "action": "history.pushQuery",
                      "args": [
                        {
                          "logId": null,
                        },
                        {
                          "notify": false,
                        },
                      ],
                    },
                  ],
                  "if": "<% EVENT.detail %>",
                  "then": [
                    {
                      "action": "context.replace",
                      "args": [
                        "logId",
                        "<% EVENT.detail %>",
                      ],
                    },
                    {
                      "action": "history.pushQuery",
                      "args": [
                        {
                          "logId": "<% EVENT.detail %>",
                        },
                        {
                          "notify": false,
                        },
                      ],
                    },
                  ],
                },
              ],
            },
            "properties": {
              "inputStyle": {
                "width": "300px",
              },
              "placeholder": "输入任务 ID 并按回车",
              "themeVariant": "elevo",
              "trim": true,
              "value": "<%= CTX.logId %>",
            },
          },
          {
            "brick": ":if",
            "children": [
              {
                "brick": "eo-button",
                "children": [
                  {
                    "brick": "span",
                    "events": undefined,
                    "properties": {
                      "textContent": "<%= CTX.allExpanded ? "全部收起" : "全部展开" %>",
                    },
                  },
                ],
                "events": {
                  "click": [
                    {
                      "action": "context.replace",
                      "args": [
                        "allExpanded",
                        "<% !CTX.allExpanded %>",
                      ],
                    },
                  ],
                },
                "properties": {
                  "themeVariant": "elevo",
                },
              },
            ],
            "dataSource": "<%= CTX.logs %>",
            "events": undefined,
          },
        ],
        "events": undefined,
        "properties": {
          "alignItems": "center",
          "gap": "0.5em",
        },
      },
      {
        "brick": ":if",
        "children": [
          {
            "brick": "eo-next-table",
            "children": [],
            "events": undefined,
            "properties": {
              "columns": [
                {
                  "dataIndex": "time",
                  "key": "time",
                  "title": "时间",
                  "useBrick": [
                    {
                      "brick": "span",
                      "events": undefined,
                      "properties": {
                        "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(moment(DATA.cellData * 1000).format("HH:mm:ss"))}) %>",
                      },
                    },
                  ],
                },
                {
                  "dataIndex": "opType",
                  "key": "opType",
                  "title": "操作类型",
                  "useBrick": [
                    {
                      "brick": "span",
                      "events": undefined,
                      "properties": {
                        "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(\`\${DATA.cellData}\${DATA.rowData.opSubType ? \` / \${DATA.rowData.opSubType}\` : ""}\`)}) %>",
                      },
                    },
                  ],
                },
                {
                  "dataIndex": "data",
                  "key": "data",
                  "title": "数据",
                  "useBrick": [
                    {
                      "brick": "div",
                      "children": [
                        {
                          "brick": ":if",
                          "children": [
                            {
                              "brick": "eo-code-block",
                              "events": undefined,
                              "properties": {
                                "language": "json",
                                "source": "<%= FN.prettierOutput(DATA.cellData) %>",
                                "style": {
                                  "maxWidth": "696px",
                                },
                                "themeVariant": "elevo",
                              },
                            },
                            {
                              "brick": "span",
                              "children": [
                                {
                                  "brick": "span",
                                  "events": undefined,
                                  "properties": {
                                    "textContent": "<%= DATA.cellData %>",
                                  },
                                },
                              ],
                              "events": undefined,
                              "properties": {
                                "style": {
                                  "display": "block",
                                  "maxWidth": "696px",
                                  "overflow": "hidden",
                                  "textOverflow": "ellipsis",
                                  "whiteSpace": "nowrap",
                                },
                              },
                              "slot": "else",
                            },
                          ],
                          "dataSource": "<%= CTX.allExpanded || CTX.expandedSet.has(DATA.rowData.logId) %>",
                          "events": undefined,
                        },
                      ],
                    },
                  ],
                },
                {
                  "dataIndex": "operation",
                  "key": "operation",
                  "title": "操作",
                  "useBrick": [
                    {
                      "brick": "isolated-tpl-0-operations",
                      "events": {
                        "open.detail": [
                          {
                            "action": "context.replace",
                            "args": [
                              "detailLog",
                              "<% DATA.rowData %>",
                            ],
                          },
                          {
                            "args": [],
                            "method": "open",
                            "target": "[data-root-id="test-root"] [data-ref="modalRef"]",
                          },
                        ],
                        "toggle": [
                          {
                            "action": "context.replace",
                            "args": [
                              "expandedSet",
                              "<% new Set(
                        CTX.expandedSet.has(DATA.rowData.logId)
                          ? [...CTX.expandedSet].filter((id) => id !== DATA.rowData.logId)
                          : [...CTX.expandedSet, DATA.rowData.logId]
                      ) %>",
                            ],
                          },
                        ],
                      },
                      "properties": {
                        "allExpanded": "<%= CTX.allExpanded %>",
                        "expanded": "<%= CTX.expandedSet.has(DATA.rowData.logId) %>",
                      },
                    },
                  ],
                },
              ],
              "dataSource": {
                "list": "<%= CTX.logs %>",
              },
              "pagination": false,
              "rowKey": "logId",
              "scrollConfig": {
                "x": "max-content",
              },
              "size": "middle",
              "themeVariant": "elevo",
            },
          },
        ],
        "dataSource": "<%= CTX.logs %>",
        "events": undefined,
      },
      {
        "brick": "eo-modal",
        "children": [
          {
            "brick": "eo-content-layout",
            "children": [
              {
                "brick": ":if",
                "children": [
                  {
                    "brick": "eo-descriptions",
                    "children": [],
                    "events": undefined,
                    "properties": {
                      "column": 1,
                      "dataSource": "<%= CTX.detailLog %>",
                      "list": [
                        {
                          "field": "logId",
                          "label": "日志 ID",
                        },
                        {
                          "field": "conversationId",
                          "label": "任务 ID",
                        },
                        {
                          "field": "taskId",
                          "label": "任务 ID",
                        },
                        {
                          "label": "时间",
                          "useBrick": [
                            {
                              "brick": "span",
                              "events": undefined,
                              "properties": {
                                "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(moment(DATA.time * 1000).format("llll"))}) %>",
                              },
                            },
                          ],
                        },
                      ],
                      "themeVariant": "elevo",
                    },
                  },
                  {
                    "brick": "eo-code-block",
                    "events": undefined,
                    "properties": {
                      "language": "json",
                      "source": "<%= FN.prettierOutput(CTX.detailLog.data) %>",
                      "themeVariant": "elevo",
                    },
                  },
                ],
                "dataSource": "<%= CTX.detailLog %>",
                "events": undefined,
              },
            ],
          },
        ],
        "events": undefined,
        "portal": true,
        "properties": {
          "cancelText": undefined,
          "confirmText": undefined,
          "dataset": {
            "ref": "modalRef",
          },
          "hideCancelButton": true,
          "keyboard": true,
          "maskClosable": true,
          "modalTitle": "<%= CTX.detailLog
            ? \`\${r.opType}\${r.opSubType ? \` / \${r.opSubType}\` : ""}\`
            : "" %>",
          "themeVariant": "elevo",
          "width": "800px",
        },
      },
    ],
  },
  "context": [
    {
      "expose": false,
      "name": "logId",
      "track": true,
      "value": "<% QUERY.logId %>",
    },
    {
      "expose": false,
      "name": "allExpanded",
      "track": true,
      "value": false,
    },
    {
      "expose": false,
      "name": "expandedSet",
      "track": true,
      "value": "<% new Set() %>",
    },
    {
      "expose": false,
      "name": "detailLog",
      "track": true,
      "value": null,
    },
    {
      "name": "logs",
      "resolve": {
        "args": [
          "<% \`logs/\${CTX.logId}\` %>",
          undefined,
        ],
        "if": "<% !!CTX.logId %>",
        "transform": {
          "value": "<% DATA.data.logs %>",
        },
        "useProvider": "basic.http-request",
      },
      "track": true,
      "value": null,
    },
  ],
  "functions": [
    {
      "name": "prettierOutput",
      "source": "function prettierOutput(output: string) {
  try {
    const data = JSON.parse(output);
    return JSON.stringify(data, null, 2);
  } catch (e) {
    return output;
  }
}",
      "typescript": true,
    },
    {
      "name": "_helper_mergeTexts",
      "source": "function mock() {}",
      "typescript": true,
    },
  ],
  "templates": [
    {
      "bricks": [
        {
          "brick": "eo-flex-layout",
          "children": [
            {
              "brick": ":if",
              "children": [
                {
                  "brick": "eo-button",
                  "events": {
                    "click": [
                      {
                        "action": "tpl.dispatchEvent",
                        "args": [
                          "toggle",
                          {
                            "detail": "<% EVENT.detail %>",
                          },
                        ],
                      },
                    ],
                  },
                  "properties": {
                    "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(STATE.expanded ? "收起" : "展开")}) %>",
                    "themeVariant": "elevo",
                  },
                },
              ],
              "dataSource": "<%= !STATE.allExpanded %>",
              "events": undefined,
            },
            {
              "brick": "eo-button",
              "events": {
                "click": [
                  {
                    "action": "tpl.dispatchEvent",
                    "args": [
                      "open.detail",
                      {
                        "detail": "<% EVENT.detail %>",
                      },
                    ],
                  },
                ],
              },
              "properties": {
                "textContent": "详情",
                "themeVariant": "elevo",
              },
            },
          ],
          "events": undefined,
          "properties": {
            "alignItems": "center",
            "gap": "0.5em",
          },
        },
      ],
      "name": "isolated-tpl-0-operations",
      "state": [
        {
          "expose": true,
          "name": "expanded",
          "track": true,
          "value": undefined,
        },
        {
          "expose": true,
          "name": "allExpanded",
          "track": true,
          "value": undefined,
        },
      ],
    },
  ],
  "title": "任务日志",
}
`;
